src/Interpreters/convertFieldToType.cpp

```diff
@@ -251,6 +251,16 @@ Field convertFieldToTypeImpl(const Field & src, const IDataType & type, const ID
 
         if (which_type.isDateTime64() && src.getType() == Field::Types::Decimal64)
         {
+            const int scale_diff = static_cast<const DataTypeDateTime64 &>(type).getScale() - src.get<Decimal64>().getScale();
+            /// in case if we need to make DateTime64(a) from DateTime64(b), we need to convert datetime value to the right scale
+            if (scale_diff != 0)
+            {
+                const auto & date_time64_type = static_cast<const DataTypeDateTime64 &>(type);
+                const UInt64 value = scale_diff > 0 ? src.get<Decimal64>().getValue().value * UInt64(pow(10, scale_diff)) : src.get<Decimal64>().getValue().value / UInt64(pow(10, -scale_diff));
```

Comment:
Please use [`scaleMultiplier`](https://clickhouse.com/codebrowser/ClickHouse/src/Core/DecimalFunctions.h.html#_ZN2DB12DecimalUtils15scaleMultiplierEj) instead of `pow`.

For reference:
https://clickhouse.com/codebrowser/ClickHouse/src/Functions/FunctionUnixTimestamp64.h.html#69